### ISSUE TITLE
off_axis_projection can't plot fields which need field parameters

### ISSUE THREAD

**Issue by jzuhone** [yt-project/yt#3126]
### Bug report

**Bug summary**

When one attempts to off-axis project a field that needs a field parameter (in this case, `"gas", "velocity_los"`), you end up with a crash because when calculating the field bounds you create a new object with `ds.all_data()` that doesn't know about the field parameter: 

**Code for reproduction**

```python
import yt
ds = yt.load("GasSloshing/sloshing_nomag2_hdf5_plt_cnt_0150")
prj = yt.OffAxisProjectionPlot(ds, [0, 0, 1], "velocity_los", weight_field="density")
```

**Actual outcome**

```
NeedsParameter: (['axis'])
---------------------------------------------------------------------------
NeedsParameter                            Traceback (most recent call last)
<ipython-input-3-57190e41cb45> in <module>
----> 1 prj = yt.OffAxisProjectionPlot(ds, [0, 0, 1], "velocity_los", weight_field="density")
~/Source/yt/yt/visualization/plot_window.py in __init__(self, ds, normal, fields, center, width, depth, axes_unit, weight_field, max_level, north_vector, right_handed, volume, no_ghost, le, re, interpolated, fontsize, method, data_source, buff_size)
   2112         # Hard-coding the origin keyword since the other two options
   2113         # aren't well-defined for off-axis data objects
-> 2114         PWViewerMPL.__init__(
   2115             self,
   2116             OffAxisProj,
~/Source/yt/yt/visualization/plot_window.py in __init__(self, *args, **kwargs)
    803             self._plot_type = kwargs.pop("plot_type")
    804         self._splat_color = kwargs.pop("splat_color", None)
--> 805         PlotWindow.__init__(self, *args, **kwargs)
    806 
    807     def _setup_origin(self):
~/Source/yt/yt/visualization/plot_window.py in __init__(self, data_source, bounds, buff_size, antialias, periodic, origin, oblique, right_handed, window_size, fields, fontsize, aspect, setup)
    230                 self._field_transform[field] = linear_transform
    231         self.setup_callbacks()
--> 232         self._setup_plots()
    233 
    234     def __iter__(self):
~/Source/yt/yt/visualization/plot_window.py in _setup_plots(self)
    897             return
    898         if not self._data_valid:
--> 899             self._recreate_frb()
    900             self._data_valid = True
    901         self._colorbar_valid = True
~/Source/yt/yt/visualization/plot_window.py in _recreate_frb(self)
    290         # At this point the frb has the valid bounds, size, aliasing, etc.
    291         if old_fields is None:
--> 292             self._frb._get_data_source_fields()
    293         else:
    294             # Restore the old fields
~/Source/yt/yt/visualization/fixed_resolution.py in _get_data_source_fields(self)
    175         for f in fields:
    176             if f not in exclude and f[0] not in self.data_source.ds.particle_types:
--> 177                 self[f]
    178 
    179     def _get_info(self, item):
~/Source/yt/yt/visualization/fixed_resolution.py in __getitem__(self, item)
    578             )
    579         )
--> 580         buff = off_axis_projection(
    581             dd.dd,
    582             dd.center,
~/Source/yt/yt/visualization/volume_rendering/off_axis_projection.py in off_axis_projection(data_source, center, normal_vector, width, resolution, item, weight, volume, no_ghost, interpolated, north_vector, num_threads, method)
    327         deps, _ = data_source.ds.field_info.check_derived_fields([weightfield])
    328         data_source.ds.field_dependencies.update(deps)
--> 329         vol.set_field(weightfield)
    330         vol.set_weight_field(weight)
    331     ptf = ProjectionTransferFunction()
~/Source/yt/yt/visualization/volume_rendering/render_source.py in set_field(self, field)
    360             The field to render
    361         """
--> 362         self.field = field
    363         return self
    364 
~/Source/yt/yt/visualization/volume_rendering/render_source.py in wrapper(*args, **kwargs)
     55         ret = f(*args, **kwargs)
     56         obj = args[0]
---> 57         if isinstance(obj.transfer_function, ProjectionTransferFunction):
     58             obj.sampler_type = "projection"
     59             obj._log_field = False
~/Source/yt/yt/visualization/volume_rendering/render_source.py in transfer_function(self)
    222         self.tfh.set_field(self.field)
    223         self.tfh.set_log(self.log_field)
--> 224         self.tfh.build_transfer_function()
    225         self.tfh.setup_default()
    226         self._transfer_function = self.tfh.tf
~/Source/yt/yt/visualization/volume_rendering/transfer_function_helper.py in build_transfer_function(self)
    108                 "Set the TransferFunctionHelper.bounds to avoid this."
    109             )
--> 110             self.set_bounds()
    111 
    112         if self.log:
~/Source/yt/yt/visualization/volume_rendering/transfer_function_helper.py in set_bounds(self, bounds)
     51         """
     52         if bounds is None:
---> 53             bounds = self.ds.all_data().quantities["Extrema"](self.field, non_zero=True)
     54             bounds = [b.ndarray_view() for b in bounds]
     55         self.bounds = bounds
~/Source/yt/yt/data_objects/derived_quantities.py in __call__(self, fields, non_zero)
    590     def __call__(self, fields, non_zero=False):
    591         fields = list(iter_fields(fields))
--> 592         rv = super().__call__(fields, non_zero)
    593         if len(rv) == 1:
    594             rv = rv[0]
~/Source/yt/yt/data_objects/derived_quantities.py in __call__(self, *args, **kwargs)
     53         storage = {}
     54         for sto, ds in parallel_objects(chunks, -1, storage=storage):
---> 55             sto.result = self.process_chunk(ds, *args, **kwargs)
     56         # Now storage will have everything, and will be done via pickling, so
     57         # the units will be preserved.  (Credit to Nathan for this
~/Source/yt/yt/data_objects/derived_quantities.py in process_chunk(self, data, fields, non_zero)
    599         for field in fields:
    600             field = data._determine_fields(field)[0]
--> 601             fd = data[field]
    602             if non_zero:
    603                 fd = fd[fd > 0.0]
~/Source/yt/yt/data_objects/data_containers.py in __getitem__(self, key)
    246                 return self.field_data[f]
    247             else:
--> 248                 self.get_data(f)
    249         # fi.units is the unit expression string. We depend on the registry
    250         # hanging off the dataset to define this unit object.
~/Source/yt/yt/data_objects/selection_objects/data_selection_objects.py in get_data(self, fields)
    209 
    210         fields_to_generate += gen_fluids + gen_particles
--> 211         self._generate_fields(fields_to_generate)
    212         for field in list(self.field_data.keys()):
    213             if field not in ofields:
~/Source/yt/yt/data_objects/selection_objects/data_selection_objects.py in _generate_fields(self, fields_to_generate)
    230                 fi = self.ds._get_field_info(*field)
    231                 try:
--> 232                     fd = self._generate_field(field)
    233                     if hasattr(fd, "units"):
    234                         fd.units.registry = self.ds.unit_registry
~/Source/yt/yt/data_objects/data_containers.py in _generate_field(self, field)
    287                 tr = self._generate_particle_field(field)
    288             else:
--> 289                 tr = self._generate_fluid_field(field)
    290             if tr is None:
    291                 raise YTCouldNotGenerateField(field, self.ds)
~/Source/yt/yt/data_objects/data_containers.py in _generate_fluid_field(self, field)
    306             rv = self._generate_spatial_fluid(field, ngt_exception.ghost_zones)
    307         else:
--> 308             rv = finfo(gen_obj)
    309         return rv
    310 
~/Source/yt/yt/fields/derived_field.py in __call__(self, data)
    275             )
    276         with self.unit_registry(data):
--> 277             dd = self._function(self, data)
    278         for field_name in data.keys():
    279             if field_name not in original_fields:
~/Source/yt/yt/visualization/volume_rendering/off_axis_projection.py in temp_weightfield(a, b)
    315         def _make_wf(f, w):
    316             def temp_weightfield(a, b):
--> 317                 tr = b[f].astype("float64") * b[w]
    318                 return tr.d
    319 
~/Source/yt/yt/data_objects/data_containers.py in __getitem__(self, key)
    246                 return self.field_data[f]
    247             else:
--> 248                 self.get_data(f)
    249         # fi.units is the unit expression string. We depend on the registry
    250         # hanging off the dataset to define this unit object.
~/Source/yt/yt/data_objects/selection_objects/data_selection_objects.py in get_data(self, fields)
    167             finfo = self.ds._get_field_info(*field)
    168             try:
--> 169                 finfo.check_available(self)
    170             except NeedsGridType:
    171                 fields_to_generate.append(field)
~/Source/yt/yt/fields/derived_field.py in check_available(self, data)
    222         """
    223         for validator in self.validators:
--> 224             validator(data)
    225         # If we don't get an exception, we're good to go
    226         return True
~/Source/yt/yt/fields/derived_field.py in __call__(self, data)
    481                 doesnt_have.append(p)
    482         if len(doesnt_have) > 0:
--> 483             raise NeedsParameter(doesnt_have)
    484         return True
    485 
NeedsParameter: (['axis'])
```

**Expected outcome**

My recollection is that we never needed to get to the `TransferFunctionHelper` before when making an `off_axis_projection`, since the `ProjectionTransferFunction` was already set. I tried to figure out an easy fix but wasn't able to. I think @cphyc worked on this recently? 

**Version Information**

  * Operating System: macOS, Linux
  * Python Version: Python 3.8, 3.9
  * yt version: 4.0-dev (9799034a7427538748f3)
  * Other Libraries (if applicable):

**Comment by jzuhone** [comment]
Note that the `"velocity_los"` field already has a field parameter set, it doesn't have to be set manually or passed in.

**Comment by jzuhone** [comment]
Here's what I've determined so far. So the relevant block of code in `off_axis_projection` is this:

https://github.com/yt-project/yt/blob/bc33f4e188f9b29a55588a4f9f7e88d65a81f2cd/yt/visualization/volume_rendering/off_axis_projection.py#L307-L332

We create a `KDTreeVolumeSource` before figuring out the fields. Then in either line 310 or line 329, we call the `set_field` on the volume. `set_field` destroys the `transfer_function` attribute by setting it to `None`, but we haven't created one yet anyway. However, if we had created a `ProjectionTransferFunction` (which happens later in lines 331-332), it would have been eliminated at this point. 

`set_field` is wrapped by `invalidate_volume`:

https://github.com/yt-project/yt/blob/bc33f4e188f9b29a55588a4f9f7e88d65a81f2cd/yt/visualization/volume_rendering/render_source.py#L52-L65

which in line 57 checks the `transfer_function` attribute, which is actually a property, so it tries to create a new one using the `TransferFunctionHelper` in line 224 here:

https://github.com/yt-project/yt/blob/bc33f4e188f9b29a55588a4f9f7e88d65a81f2cd/yt/visualization/volume_rendering/render_source.py#L211-L228

But in there, it tries to figure out the bounds of the field using a call to `ds.all_data()` in line 53 here, which knows nothing about the field parameters, resulting in the crash:

https://github.com/yt-project/yt/blob/bc33f4e188f9b29a55588a4f9f7e88d65a81f2cd/yt/visualization/volume_rendering/transfer_function_helper.py#L40-L62

So I get why this is happening now, but I have no idea how to fix it.

**Comment by jzuhone** [comment]
@matthewturk @chummels @cphyc @neutrinoceros the above comment ^^ is my analysis of what's happening here--all of you have worked in this part of the code at some point, so any advice on how to fix this would be great. See also, which is where the issue arose:

https://github.com/yt-project/yt/commit/cde25a88935c5a0a7b6fa17602bb27b7f2e8a864#diff-2d1d24ee7df2e3b516614aad6fa0d7bd09a0268684a1c6b748eee75c1b823002

The problem is that I think @cphyc made the right call in that change in line 57.

**Comment by neutrinoceros** [comment]
I'm really sorry I don't have any memory of working on this :/

**Comment by jzuhone** [comment]
@neutrinoceros no worries

**Comment by munkm** [comment]
This is closed via yt-project/yt#3147. 

If I'm mistaken, feel free to reopen!
